### ISSUE TITLE
Consider use of SharedArrayBuffer and associated web technologies

### ISSUE THREAD

**Issue by filiph** [flutter/flutter#153760]
### Use case

I was looking at game engines like Unity and Godot recently, and I noticed that when these engines target the web, they _require_ some technologies, such as `SharedArrayBuffer` (and the associated Cross Origin Isolation). From what I understand, these technologies enable efficient WASM multithreading. (A good primer on this tech is [in this Godot Engine article](https://godotengine.org/article/progress-report-web-export-in-4-3/#single-threaded-web-export).)

Requiring things like `SharedArrayBuffer` brings a problem: they are disabled by default on most servers, and they will probably stay disabled. The thing is, when SharedArrayBuffer is enabled on a page, some pretty basic web things stop working. For example, embedding (such as embedding a YouTube video or an ad). Also, even if you decide to enable it, it's non-trivial on localhost (needs SSH), and hard to impossible on some popular hosting platforms (e.g. GitHub Pages doesn't have that option).

So, on one hand, the fact that Flutter _isn't_ using `SharedArrayBuffer` is great, because you can have Flutter apps and games running on any page, using almost any server config. This could even be a competitive advantage in the casual web game space.

On the other hand, `SharedArrayBuffer` and multithreading might be a low-hanging fruit for optimization that Flutter is currently not using. There's probably a reason why both Unity and Godot require it.

// cc @zoeyfan 

(For context, here are two previous issues mentioning SharedArrayBuffer: https://github.com/flutter/flutter/issues/126351 and https://github.com/flutter/flutter/issues/124334.)

### Proposal

As a developer, I'd like to have the ability to make my Flutter apps/games go faster on the web. Maybe there can be a flag that enables these technologies, so I can make the trade-off decision myself. 

If the Flutter team decides not to explore `SharedArrayBuffer` et al. for the time being, then it would be good to know. At the very least, for games there's the competitive advantage of running anywhere on the web (in contrast to Unity and, [to a lesser extent](https://godotengine.org/article/progress-report-web-export-in-4-3/), Godot), and without the need to configure the server.

**Comment by filiph** [comment]
Do I understand it correctly that there's auto-detection going on at runtime? That's amazing!

In that case, I think it would be great to point this out more explicitly, and especially in the context of performance on the web. I'm not sure what the performance upside can be when using multi-threading, but assuming the upside is non-trivial, I think it's worth explicitly teaching people how to enable `SharedArrayBuffer` on their hosting platform (example [here](https://github.com/godotengine/godot/issues/69020#issuecomment-1383867679)).

On Firebase hosting and on itch.io, for instance, it's not that hard to enable `SharedArrayBuffer`. But one needs to first know that this is desirable.

I'd say this is a DevRel / `docs` issue then?

**Comment by yjbanov** [comment]
@filiph You are right, the benefits in Flutter are significant with multi-threading, and yes, the security restrictions are steep for some apps. The web has been built around fetching arbitrary resources from arbitrary origins, so it's not always easy to take an existing app and restrict it to allow shared memory multithreading. The best approach today would be to start and evolve the app under those restrictions. Games should be particularly conducive for this, as the resources they use are typically bundled with the app. There's not much cross-origin stuff going on, and whatever cross-origin functionality is needed, it can probably be accommodated using an iframe (e.g. authentication).

Our current fallback is pretty drastic. If there's no `SharedArrayBuffer` we fallback all the way to JS and CanvasKit. We are going to add a softer fallback, where you still get to use dart2wasm but render in single-threaded mode (https://github.com/flutter/flutter/issues/126351). This should work for most apps out-of-the-box, with no changes to cross-origin functionality.

We are also watching how the web ecosystem is evolving. It's possible the web will figure out a way to allow multi-threaded rendering without the extra restrictions. We're also casually discussing ideas for doing multi-threaded rendering without the `SharedArrayBuffer`, e.g. by using [transferable objects](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects) (this would be tricky, as Skia is designed around shared memory multi-threading, e.g. an `SkPaint` object is needed both on the UI thread and on the GPU thread).

**Comment by yjbanov** [comment]
I'm inclined to close this issue as it seems everything is covered by existing issues, but LMK if there's something specific we can do.

**Comment by filiph** [comment]
I'd like to pivot this issue to documentation, then. 

- Should we add a note to https://docs.flutter.dev/perf/rendering-performance#web-only-advice, for instance? 
- Elsewhere in the `/perf/*` section of the site?
- Should we write a cookbook recipe in https://docs.flutter.dev/cookbook/games ?
- Should we expand on the paragraph in https://docs.flutter.dev/platform-integration/web/renderers#skwasm ?

// cc @zoeyfan

**Comment by yjbanov** [comment]
Agreed. Docs would be great!
